### PR TITLE
Multiple improvements for stm32wle targets

### PR DESCRIPTION
**Pull request by ofauchon**
This PR provide multiple improvements for STM32WLE5 targets:

- board/stm32: Add STM32 Nucleo WL55JC board
- board/lorae5,nucleo-wl55jc: Add SubGhz SPI3 interface
- stm32/stm32wl: Set 48Mhz HSE32/PLL as default Clock, SubGhz Radio device init

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -115,6 +115,7 @@ The following 72 microcontroller boards are currently supported:
 * [ST Micro "Nucleo" L031K6](https://www.st.com/ja/evaluation-tools/nucleo-l031k6.html)
 * [ST Micro "Nucleo" L432KC](https://www.st.com/ja/evaluation-tools/nucleo-l432kc.html)
 * [ST Micro "Nucleo" L552ZE](https://www.st.com/en/evaluation-tools/nucleo-l552ze-q.html)
+* [ST Micro "Nucleo" WL55JC](https://www.st.com/en/evaluation-tools/nucleo-wl55jc.html)
```
Comment (kenbell) [inline README.md:118]: Should increment the count of supported boards in the text above

File: src/runtime/runtime_stm32wle5.go [hunk 2]
```diff
@@ -4,21 +4,125 @@
...
 	initTickTimer(&machine.TIM1)
+
+	// Radio init
+	subGhzInit()
+}
+
+func initCLK() {
+
+	// Set Power Voltage Regulator Range 2
+	//stm32.PWR.CR1.ReplaceBits(0b10, stm32.PWR_CR1_VOS_Msk, stm32.PWR_CR1_VOS_Pos)
```
Comment (kenbell) [inline src/runtime/runtime_stm32wle5.go]: Should either un-comment, or completely remove

File: src/runtime/runtime_stm32wle5.go [hunk 3]
```diff
@@ -4,21 +4,125 @@
...
+	for !stm32.RCC.CR.HasBits(stm32.RCC_CR_PLLRDY) {
+	}
+
+	// Enable PLL System Clock output.
+	stm32.RCC.PLLCFGR.SetBits(stm32.RCC_PLLCFGR_PLLREN)
+	for !stm32.RCC.CR.HasBits(stm32.RCC_CR_PLLRDY) {
+	}
+
+	// Set Flash Latency of 2 and wait until it's set properly
+	stm32.FLASH.ACR.ReplaceBits(0b010, 0b111, stm32.Flash_ACR_LATENCY_Pos)
```
Comment (kenbell) [inline src/runtime/runtime_stm32wle5.go]: Can use symbolic constants for these values?

File: src/runtime/runtime_stm32wle5.go [hunk 4]
```diff
@@ -4,21 +4,125 @@
...
+	for !stm32.RCC.CR.HasBits(stm32.RCC_CR_HSERDY) {
+	}
+
+	// Disable PLL
+	stm32.RCC.CR.ClearBits(stm32.RCC_CR_PLLON)
+	for stm32.RCC.CR.HasBits(stm32.RCC_CR_PLLRDY) {
+	}
+
+	// Configure PLL
+	stm32.RCC.PLLCFGR.Set(0x23020613)
```
Comment (kenbell) [inline src/runtime/runtime_stm32wle5.go]: Can use symbolic constants here?  E.g. other STM32 worked on recently tend to have something like this:
```
	// Configure the PLL
	stm32.RCC.PLLCFGR.ReplaceBits(
		(1)| // 1 = RCC_PLLSOURCE_MSI
			(PLL_M-1)<<stm32.RCC_PLLCFGR_PLLM_Pos|
			(PLL_N<<stm32.RCC_PLLCFGR_PLLN_Pos)|
			(((PLL_Q>>1)-1)<<stm32.RCC_PLLCFGR_PLLQ_Pos)|
			(((PLL_R>>1)-1)<<stm32.RCC_PLLCFGR_PLLR_Pos)|
			(PLL_P<<stm32.RCC_PLLCFGR_PLLPDIV_Pos),
		stm32.RCC_PLLCFGR_PLLSRC_Msk|stm32.RCC_PLLCFGR_PLLM_Msk|
			stm32.RCC_PLLCFGR_PLLN_Msk|stm32.RCC_PLLCFGR_PLLP_Msk|
			stm32.RCC_PLLCFGR_PLLR_Msk|stm32.RCC_PLLCFGR_PLLPDIV_Msk,
		0)
```

File: src/runtime/runtime_stm32wle5.go [hunk 5]
```diff
@@ -4,21 +4,125 @@
...
+	}
+
+	// Set clock source to PLL (0x3)
+	stm32.RCC.CFGR.ReplaceBits(0b11, 0b11, stm32.RCC_CFGR_SW_Pos)
+	for (stm32.RCC.CFGR.Get() & stm32.RCC_CFGR_SWS_Msk) != 0xc {
+	}
+
+}
+
+// SubGhzInit() enables SX126x radio module
```
Comment (kenbell) [inline src/runtime/runtime_stm32wle5.go]: I don't think this should be in the runtime package - runtime is really about setting up clocks and time.

Should probably think about this as an optional peripheral that apps may (or may not) use - with it either:
1 - being in machine package
2 - being supported in the tinygo drivers repo

I'm not sure which is best - it's not a SPI/I2C peripheral - and might be a bit 'specialized' for the machine package?  Perhaps @aykevl  or @deadprogram  have some thoughts?

**Review by kenbell — COMMENTED** [review]
Thanks for improving the support - suggest just a small amount of replacing magic values with symbolics (where possible) - and needs some thought about how (and when) to initialize the radio peripheral.

Also - looks like a change may be missing based on the compile errors?

File: src/runtime/runtime_stm32wle5.go (hunk 5 above)
Comment (ofauchon) [reply src/runtime/runtime_stm32wle5.go]: SX126x Radio (which is part of STM32WL) will have its own tinygo driver soon, so the code can be reused when using SX126x standalone module.
 
But this function is specific to STM32WL series (as SX126x radio unit and MCU are on the same die)
 
IMHO it's entirely part of the whole chipset initialisation process.

I'm not sure moving this function to tinygo-driver makes sense (I don't see other situations you might need this)

But having it in machine package instead, would not hurt. 

Other comments welcome .

Comment (deadprogram) [reply src/runtime/runtime_stm32wle5.go]: `machine` package is a good place for this. Perhaps it should be exported so users who want to turn on this functionality can do so, but it is not always turned on by default?

**Comment by deadprogram** [comment]
Once you have updated from any changes from feedback, @ofauchon can you please rebase against the latest `dev` branch? Hopefully that will get CI to build successfully.

**Comment by deadprogram** [comment]
Something else must be wrong here @ofauchon see https://app.circleci.com/pipelines/github/tinygo-org/tinygo/6087/workflows/7a6938bf-d87d-4690-8782-209e919ae383/jobs/34339?invite=true#step-117-360

File: src/runtime/runtime_stm32wle5.go [hunk 6]
```diff
@@ -4,21 +4,125 @@
...
+	initCLK()
 
 	// UART init
 	machine.Serial.Configure(machine.UARTConfig{})
 
 	// Timers init
 	initTickTimer(&machine.TIM1)
+
+	// Radio init
+	subGhzInit()
```
Comment (deadprogram) [inline src/runtime/runtime_stm32wle5.go]: CI is now passing @ofauchon will you move `subGhzInit()` to `machine` package now, change to `SubGhzInit()`, and remove this call here so initialization is optional to the developer? :smile_cat:

File: src/runtime/runtime_stm32wle5.go (hunk 2 above)
Comment (ofauchon) [reply src/runtime/runtime_stm32wle5.go]: Removed

File: src/runtime/runtime_stm32wle5.go (hunk 3 above)
Comment (ofauchon) [reply src/runtime/runtime_stm32wle5.go]: All binaries replaced with symbolic constants.

File: src/runtime/runtime_stm32wle5.go (hunk 4 above)
Comment (ofauchon) [reply src/runtime/runtime_stm32wle5.go]: I could rewrite the PLL init with symbolic constants . Done

File: src/runtime/runtime_stm32wle5.go (hunk 5 above)
Comment (ofauchon) [reply src/runtime/runtime_stm32wle5.go]: SubGhzInit function moved in "src/machine/machine_stm32wle5.go" and exported so users can use it or not

**Comment by ofauchon** [comment]
@deadprogram @aykevl @kenbell . 
I made all the suggested changes. 

The final PR Code  was tested on both LoraE5 and nucleo-wl55jc: 

- Simple timers tests OK , 
- UART OK, 
- Radio module SX126x simple tests (Continuous TX checked with a SDR)

Olivier

File: src/machine/board_lorae5.go [hunk 7]
```diff
@@ -9,8 +9,8 @@ import (
 )
 
 const (
-	LED1 = PB5
-	LED  = LED1 // Default LED
+	// We assume a LED is connected on PB5
+	LED = PB5 // Default LED
```
Comment (aykevl) [inline src/machine/board_lorae5.go:13]: It looks like this change is unrelated to the PR. Was it intentional?

File: src/machine/machine_stm32wle5.go [hunk 8]
```diff
@@ -221,6 +225,61 @@ func (p Pin) registerInterrupt() interrupt.Interrupt {
...
+}
+
+func (spi SPI) getBaudRate(config SPIConfig) uint32 {
+	var clock uint32
+	switch spi.Bus {
+	case stm32.SPI1:
+		clock = CPUFrequency()
+	case stm32.SPI2, stm32.SPI3:
+		clock = CPUFrequency()
+	}
```
Comment (aykevl) [inline src/machine/machine_stm32wle5.go:257]: I don't understand this part. In all cases the clock equals the CPU frequency? Why not write `clock := CPUFrequency()` instead?

**Review by aykevl — COMMENTED** [review]
Generally looks good to me.
I'm a little bit unsure about the `SubGhzInit` function. IIRC it's on the same chip, but connected through an (on-chip) SPI bus?
I'm almost inclined to say it should be implemented entirely in a driver package, not in the machine package. Not sure. On the other hand, it does need some special initialization so it's not just any other SPI peripheral.

File: src/machine/board_lorae5.go (hunk 7 above)
Comment (ofauchon) [reply src/machine/board_lorae5.go:13]: This board don't actually have a LED. 
So I thought I would be a good idea to remove it. 

But, that breaks blinky1 , so I rolled-back the change.

File: src/machine/machine_stm32wle5.go (hunk 8 above)
Comment (ofauchon) [reply src/machine/machine_stm32wle5.go:257]: IIRC, this code was (quickly) borrowed from STM32F405

I guess the lines are the same because the both SPI1 and SPI2/SPI3 's bus use a 1:1 Prescaler. 
(in the current STM32WL setup) 

In the future, we may improve / tune the clocks, and the lines may turn into :

```
	switch spi.Bus {
	case stm32.SPI1:
		clock = CPUFrequency() /2 
	case stm32.SPI2, stm32.SPI3:
		clock = CPUFrequency() /4
	}
```
So we have two solutions: 

  - Add comments to explain the PRESCALERs are 1:1 in the current configuration
  - Review and complete the work on SYSCLK/BUS/SPI clocks , prescalers... but that will take time
 
As I plan to work on Power Saving for this chip in a near future, 
i'd prefer not spending too much time on the clocks right now, and juste add
some comment. 

Tell me what you feel the most confortable with . 

Olivier

Comment (deadprogram) [reply src/machine/machine_stm32wle5.go:257]: Add comment about keeping separate to support future improvements, and IMO its all good.

Comment (ofauchon) [reply src/machine/machine_stm32wle5.go:257]: I added the comment. 
Thank you all for the time spend on this PR !

**Comment by ofauchon** [comment]
> Generally looks good to me. I'm a little bit unsure about the `SubGhzInit` function. IIRC it's on the same chip, but connected through an (on-chip) SPI bus? I'm almost inclined to say it should be implemented entirely in a driver package, not in the machine package. Not sure. On the other hand, it does need some special initialization so it's not just any other SPI peripheral.

Yes, SPI3/SubGhz Is an internal SPI bus dedicated to SX126X Radio Submodule communication.

I agree this SubGhzInit is not ideal it looks like an ugly quirk in machine code.

I'll remove this function for the moment, and try to implement this in the tinygo-driver.

File: src/machine/board_lorae5.go (hunk 7 above)
Comment (kenbell) [reply src/machine/board_lorae5.go:13]: FWIW in the LoRa-e5 schematic, PB5 is connected to a RED LED through a jumper (see bottom right of the schematic): https://files.seeedstudio.com/products/113990934/LoRa-E5%20Dev%20Board%20v1.0.pdf

It certainly blinked the LED for me :)

Comment (ofauchon) [reply src/machine/board_lorae5.go:13]: image

I understand why i've never seen this led blinking and why I connected an external LED (hopefully on PB5)

Comment (kenbell) [reply src/machine/board_lorae5.go:13]: :-) the official dev board does look a little less artisanal.

**Comment by deadprogram** [comment]
Thanks for responding to all the feedback @ofauchon 

Now merging.
